### PR TITLE
Create subset_sum.cpp

### PR DESCRIPTION
**Pull request by Swastyy**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1517"></a>

### CODE REVIEW

**Comment by Swastyy** [comment]
@fhlasek I have added self test cases. Please have a look.

P.S. It returns the number of subsets with given sum since storing all the subsets and then returning them was taking much more space leading to bad space complexity so I preferred returning an integer.

File: backtracking/subset_sum.cpp
Comment (fhlasek) [inline backtracking/subset_sum.cpp:38]: Suggestion: you actually don't even need to maintain the subset since the only thing you use is the sum.

File: backtracking/subset_sum.cpp
Comment (fhlasek) [inline backtracking/subset_sum.cpp:71]: nit: this repeating comment is a bit redundant here.

File: backtracking/subset_sum.cpp
Comment (fhlasek) [inline backtracking/subset_sum.cpp:58]: An interesting corner case might be an empty array.

File: backtracking/subset_sum.cpp
Comment (fhlasek) [inline backtracking/subset_sum.cpp:85]: A more useful comment might be explaining the correct answer.

File: backtracking/subset_sum.cpp
Comment (fhlasek) [inline backtracking/subset_sum.cpp:20]: Would be useful to describe the algorithm and its complexity.

**Review by fhlasek — CHANGES_REQUESTED** [review]
Nice, simple, clear and well documented and tested. Thanks!

File: backtracking/subset_sum.cpp
Comment (Swastyy) [reply backtracking/subset_sum.cpp:38]: Yep, no need seriously 😄

File: backtracking/subset_sum.cpp
Comment (Swastyy) [reply backtracking/subset_sum.cpp:71]: Removed!

File: backtracking/subset_sum.cpp
Comment (Swastyy) [reply backtracking/subset_sum.cpp:58]: 💯 Added

File: backtracking/subset_sum.cpp
Comment (Swastyy) [reply backtracking/subset_sum.cpp:85]: 👍

File: backtracking/subset_sum.cpp
Comment (Swastyy) [reply backtracking/subset_sum.cpp:20]: Done

**Comment by Swastyy** [comment]
@fhlasek done the changes. Pl have a look.

File: backtracking/subset_sum.cpp
Comment (fhlasek) [inline backtracking/subset_sum.cpp:21]: This is not quite true. The complexity is going to be at least exponential (and actually independent on the sum.
```suggestion
 * Time Complexity: O(n * 2^n), where ‘n’ is the number of elements in the given array.
```

File: backtracking/subset_sum.cpp
Comment (fhlasek) [inline backtracking/subset_sum.cpp:33]: The comment is wrong and unnecessary
```suggestion
    int nelement = in_arr.size();
```

**Review by fhlasek — APPROVED** [review]
Code looks great now! Happy to approve. Consider some suggestions about the documentation. A better description of the algorithm might be helpful.

**Comment by Swastyy** [comment]
> Code looks great now! Happy to approve. Consider some suggestions about the documentation. A better description of the algorithm might be helpful.

I just commited the suggestions, then why there is statement like "Swastyy dismissed fhlasek’s stale review via a32b32f 3 minutes ago" in the thread?

**Comment by Swastyy** [comment]
> > Code looks great now! Happy to approve. Consider some suggestions about the documentation. A better description of the algorithm might be helpful.
> 
> I just commited the suggestions, then why there is statement like "Swastyy dismissed fhlasek’s stale review via [a32b32f](https://github.com/TheAlgorithms/C-Plus-Plus/commit/a32b32f01c3f5fd318d6de21ed0de73fe6401a81) 3 minutes ago" in the thread?

Also I can't see the approved tick. Did I made some mistake while committing suggestions?

P.S. I am newbie in open source and I don't know how these silly mistakes happens 😅

**Review by fhlasek — APPROVED** [review]
No worries, I guess that was my mistakes. I haven't realised that this repo is set up to invalidate the approval after code changes are committed. Here is another ✅ 👍🏼

File: backtracking/subset_sum.cpp
Comment (fhlasek) [inline backtracking/subset_sum.cpp:22]:
```suggestion
 * array.
 ```

**Comment by Swastyy** [comment]
@fhlasek how to get 1 more approvel? 😅

**Review by fhlasek — APPROVED** [review]
💯

**Comment by Swastyy** [comment]
@mishraabhinn Can you see this PR. It already got 1 approval, 1 more will be appreciated 😃

**Review by mishraabhinn — COMMENTED** [review]
What should be the sum =2 and arr=[1,1,1] @Swastyy

**Comment by mishraabhinn** [comment]
> @mishraabhinn Can you see this PR. It already got 1 approval, 1 more will be appreciated 😃

I think you can add a specific description, subset_sum can cause ambiguity that whether this algorithm is for  continuous subarrays or non  continuous subarrays. 
For the sum =2 and arr =[1,1,1] if former case is taken then outut will be 2 and if latter is taken than 3.

**Comment by Swastyy** [comment]
> What should be output if the sum =2 and arr=[1,1,1] @Swastyy

Yea, it will return 3. I have taken a similar test case (Test 4) with arr = [3,3,3,3] and sum = 6

**Comment by Swastyy** [comment]
> > @mishraabhinn Can you see this PR. It already got 1 approval, 1 more will be appreciated 😃
> 
> I think you can add a specific description, subset_sum can cause ambiguity that whether this algorithm is for continuous subarrays or non continuous subarrays.
> For the sum =2 and arr =[1,1,1] if former case is taken then outut will be 2 and if latter is taken than 3.

Generally subsets contain both the continuous and non-continuous subarrays. I will add it in the description. 👍

File: backtracking/subset_sum.cpp
Comment (mishraabhinn) [inline backtracking/subset_sum.cpp:4]:
```suggestion
 * @brief Program to Count number of subsets non-continuous
 * subarray with a given sum
```
Your code will fail if only continuous array is allowed.

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

**Comment by Swastyy** [comment]
We can make another file as subarray-sum.cpp which will give only the number of continuous subsets with given sum as o/p since our current file is for the total number of subarrays both continuous and noncontinuous. Let me know your thoughts @mishraabhinn

**Comment by mishraabhinn** [comment]
> We can make another file as subarray-sum.cpp which will give only the number of continuous subsets with given sum as o/p since our current file is for the total number of subarrays both continuous and noncontinuous. Let me know your thoughts @mishraabhinn

You can go ahead with this thought. And this time only continuous subarray are counted

File: backtracking/subset_sum.cpp
Comment (mishraabhinn) [inline backtracking/subset_sum.cpp:4]:
```suggestion
 * @brief We are given with an array and a sum value. The algorithms find all the subsets of that array with sum equal to given sum and return such subsets count. This approach will have exponential time complexity. 
```

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:7]:
```suggestion
 * @brief Implementation of the [Subset Sum](https://en.wikipedia.org/wiki/Subset_sum_problem) problem.
 * @details
 * We are given an array and a sum value. The algorithms find all
 * the subsets of that array with sum equal to the given sum and return such subsets
 * count. This approach will have exponential time complexity.
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:17]:
```suggestion
 * @brief Backtracking algorithms
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:24]: The Time Complexity should be above in the `details` part.
```suggestion
 * @brief Functions for the [Subset Sum](https://en.wikipedia.org/wiki/Subset_sum_problem) problem.
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:26]:
```suggestion
namespace subset_sum {
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:34]: You can use `uint64_t` instead of `std::uint64_t`.
```suggestion
uint64_t subset_sum(int sum, const std::vector<int> &in_arr) {
```
Also, remember to change the parameters to use `uint`-based types instead.

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:105]:
```suggestion
    test();  // run self-test implementations
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:8]:
```suggestion
 * @author [Swastika Gupta](https://github.com/Swastyy)
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:51]:
```suggestion
}  // namespace subset_sum
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:62]:
```suggestion
    assert(backtracking::subset_sum::subset_sum(0, array1) ==
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:70]:
```suggestion
    assert(backtracking::subset_sum::subset_sum(6, array2) ==
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:78]:
```suggestion
    assert(backtracking::subset_sum::subset_sum(1, array3) ==
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:86]:
```suggestion
    assert(backtracking::subset_sum::subset_sum(6, array4) ==
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:94]:
```suggestion
    assert(backtracking::subset_sum::subset_sum(6, array5) ==
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work.

**Comment by Panquesito7** [comment]
Glad to see more reviewers are coming back! It was quite a tough job, especially when there are like 10+ pending PRs to review. 😅

**Comment by Panquesito7** [comment]
Isn't this the same PR as #1517? I'm kind of confused...

**Comment by Swastyy** [comment]
> Isn't this the same PR as #1517? I'm kind of confused...

Did you mean #1527 ? 
No they aren't same because here we will count the overall subsets whether continuous or discontinuous with given sum.
But #1527 is only for continuous subarrays count with given sum.

File: backtracking/subset_sum.cpp
Comment (mishraabhinn) [inline backtracking/subset_sum.cpp:34]:
```suggestion
uint64_t subset_sum(uint64_t sum, const std::vector<uint64_t> &in_arr) {
```

File: backtracking/subset_sum.cpp
Comment (mishraabhinn) [inline backtracking/subset_sum.cpp:61]:
```suggestion
    std::vector<uint64_t> array1 = {-7, -3, -2, 5, 8};  // input array
```

File: backtracking/subset_sum.cpp
Comment (mishraabhinn) [inline backtracking/subset_sum.cpp:69]:
```suggestion
    std::vector<uint64_t> array2 = {1, 2, 3, 3};
```

File: backtracking/subset_sum.cpp
Comment (mishraabhinn) [inline backtracking/subset_sum.cpp:77]:
```suggestion
    std::vector<uint64_t> array3 = {1, 1, 1, 1};
```

File: backtracking/subset_sum.cpp
Comment (mishraabhinn) [inline backtracking/subset_sum.cpp:85]:
```suggestion
    std::vector<uint64_t> array4 = {3, 3, 3, 3};
```

File: backtracking/subset_sum.cpp
Comment (mishraabhinn) [inline backtracking/subset_sum.cpp:93]:
```suggestion
    std::vector<uint64_t> array5 = {};
```

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

**Comment by Swastyy** [comment]
@mishraabhinn we cannot take unsigned int as data type because we are taking negative weights array in the test case

**Comment by mishraabhinn** [comment]
> @mishraabhinn we cannot take unsigned int as data type because we are taking negative weights array in the test case

int8_t
int16_t
int32_t
int64_t
Then you should be using these specifically.

**Comment by Swastyy** [comment]
> > @mishraabhinn we cannot take unsigned int as data type because we are taking negative weights array in the test case
> 
> int8_t
> int16_t
> int32_t
> int64_t
> Then you should be using these specifically.

Now I feel like using int is not a good practice. Does it matters when we are adding our own test cases
I have changed the code accordingly. Pl see once.

**Comment by mishraabhinn** [comment]
> > > @mishraabhinn we cannot take unsigned int as data type because we are taking negative weights array in the test case
> > 
> > 
> > int8_t
> > int16_t
> > int32_t
> > int64_t
> > Then you should be using these specifically.
> 
> Now I feel like using int is not a good practice. Does it matters when we are adding our own test cases
> I have changed the code accordingly. Pl see once.

I will tell you one thing, writing code for any problem is simple. But what we should learn is to write clean code. You should write code in such a way that if you/someone someday come back to the code, you don't have to go line by line in the code to understand it, rather seeing the name of file , function name and variable name code should speak itself.

**Comment by Swastyy** [comment]
> I will tell you one thing, writing code for any problem is simple. But what we should learn is to write clean code. You should write code in such a way that if you/someone someday come back to the code, you don't have to go line by line in the code to understand it, rather seeing the name of file , function name and variable name code should speak itself.

Yea, this used to happen with me. But now I make proper doc. 😃

File: backtracking/subset_sum.cpp
Comment (mishraabhinn) [inline backtracking/subset_sum.cpp:34]: namespace is similar to function name. Function is returning count of subset. You can think of some other name.

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

**Review by mishraabhinn — APPROVED** [review]

**Comment by Swastyy** [comment]
@Panquesito7 Pl review this PR.

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:6]:
```suggestion
 * We are given an array and a sum value. The algorithm finds all
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:33]:
```suggestion
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:83]:
```suggestion
    // 4th test
```

File: backtracking/subset_sum.cpp
Comment (Panquesito7) [inline backtracking/subset_sum.cpp:13]:
```suggestion
#include <iostream>  /// for IO operations
```

**Review by Panquesito7 — APPROVED** [review]
Great work, @Swastyy! Thank you for your patience, dedication, and contributions to our community! 👍 🎉

**Pull request merged**
